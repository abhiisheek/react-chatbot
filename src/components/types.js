export default {
  TEXT: 'TEXT',
  TEXTWITHLINK: 'TEXT_WITH_LINK',
  ORDERDETAILS: 'ORDER_DETAILS',
  SUGGESTION: 'SUGGESTION',
  ORDERITEMS: 'ORDER_ITEMS',
  PIECHART: 'PIE_CHART',
  LINECHART: 'LINE_CHART',
  VBARCHART: 'V_BAR_CHART',
  HBARCHART: 'H_BAR_CHART'
};