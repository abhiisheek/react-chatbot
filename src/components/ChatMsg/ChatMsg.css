@import '../_colors.css';
@import '../_defaults.css';

.msg {
  display: flex;
}

.selfMsg {
  justify-content: flex-end;
}

.selfWrapper {
  text-align: right;
}

.chatMsgWrapper {
  padding-bottom: 5px;
}

.selfMsgWrapper {
  @apply --self-msg-wrapper;
}

.chart {
  width: 100%;
}

/* .selfMsgWrapper:before {
  @apply --msg-arrow-edge;
  right: -14px;
  border-left-color: var(--self-msg-arrow-edge-color);
} */

.botMsgWrapper {
  @apply --bot-msg-wrapper;
}

/* .botMsgWrapper:before {
  @apply --msg-arrow-edge;
  left: -14px;
  border-right-color: var(--msg-arrow-edge-color);
} */
