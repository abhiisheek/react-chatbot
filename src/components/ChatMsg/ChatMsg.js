import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import classNames from 'classnames';

import Avatar from '../Avatar';
import TimeStamp from '../Timestamp';
import ChatMsgWrapper from '../ChatMsgWrapper';

import styles from './ChatMsg.css';

import withConditionalType from '../hocs/WithConditionalType';

const ConditionalChatMsg = withConditionalType(ChatMsgWrapper);

/** Component with assembles the type component depending on the type of response along with appropriate avatar */
class ChatMsg extends PureComponent {
  rederChatMsg = ({
    self,
    selfIcon,
    showSelfAvatar,
    botIcon,
    determineTypeComp,
    ...others
  }) =>
    self ? (
      <div className={classNames(styles.chatMsgWrapper, styles.selfWrapper)}>
        <div className={classNames(styles.msg, styles.selfMsg)}>
          <ConditionalChatMsg
            {...others}
            conditionalFunc={determineTypeComp}
            className={styles.selfMsgWrapper}
          />
          {showSelfAvatar && <Avatar icon={selfIcon} />}
        </div>
        <TimeStamp />
      </div>
    ) : (
      <div className={styles.chatMsgWrapper}>
        <div className={styles.msg}>
          <Avatar icon={botIcon} />
          <ConditionalChatMsg
            {...others}
            conditionalFunc={determineTypeComp}
            className={classNames(styles.botMsgWrapper, {
              [styles.chart]: others.type.includes('CHART')
            })}
          />
        </div>
        <TimeStamp />
      </div>
    );

  render() {
    return this.rederChatMsg(this.props);
  }
}

ChatMsg.displayName = 'ChatMsg';

ChatMsg.propTypes = {
  /** Indicates whether its a self/user input response */
  self: PropTypes.bool,
  /** User's avatar icon */
  selfIcon: PropTypes.string.isRequired,
  /** Determines whether user's avatar needs to be shown are not */
  showSelfAvatar: PropTypes.bool.isRequired,
  /** Bot avatar icon */
  botIcon: PropTypes.string.isRequired,
  /** Function which determines the type of components to be used for the given response */
  determineTypeComp: PropTypes.func.isRequired,
  /** Response to be displayed */
  data: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object
  ]).isRequired
};

ChatMsg.defaultPorps = {
  self: false
};

export default ChatMsg;
