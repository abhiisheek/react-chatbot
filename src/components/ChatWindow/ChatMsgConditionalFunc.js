import PlainText from '../PlainText';
import TextWithLink from '../TextWithLink';
import Types from '../types';

const chatMsgTypesMap = {
  [Types.TEXT]: PlainText,
  [Types.TEXTWITHLINK]: TextWithLink
};

export const chatMsgConditionFunc = ({ type = Types.TEXT }) => {
  let TypeComp = chatMsgTypesMap[type];
  if (!TypeComp) {
    TypeComp = PlainText;
  }
  return TypeComp;
};
