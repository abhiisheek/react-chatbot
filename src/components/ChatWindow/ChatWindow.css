@import '../_variables.css';

.chat {
  font-family: var(--preferred-font);
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  height: 100%;
}

.chatMsgsContainer {
  overflow: auto;
  position: relative;
  margin-bottom: 5px;
  padding: 6px;
}

.chatWrapper {
  height: calc(100% - var(--header-size));
}

.docked {
  display: none;
}

.wismoIcon::before {
  content: 'WIS';
}

.categoriesWrapper {
  display: flex;
  margin-top: 40px;
  margin-bottom: 40px;
  justify-content: space-evenly;
  flex-wrap: wrap;
}

.categoryCard {
  width: 25%;
  margin: 1px 1px 40px 1px;
}
