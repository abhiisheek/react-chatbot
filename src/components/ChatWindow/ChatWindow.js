/* global setTimeout */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import classNames from 'classnames';

import ChatMsg from '../ChatMsg';
import Loader from '../Loader';
import Header from '../Header';
import TextInputForm from '../TextInputForm';
import CategoryCard from '../CategoryCard';
import { chatMsgConditionFunc } from './ChatMsgConditionalFunc';
import styles from './ChatWindow.css';

/** ChatWindow component with default config and capability to handle user input and bot responses */
class ChatWindow extends PureComponent {
  renderMessages = (
    messages,
    selfIcon,
    botIcon,
    determineTypeComp,
    showSelfAvatar,
    onSuggestionClick
  ) =>
    messages.map((message, i) => {
      const msgProp = {
        ...message,
        selfIcon,
        botIcon,
        determineTypeComp,
        showSelfAvatar
      };
      return (
        <ChatMsg {...msgProp} key={i} onClickHandler={onSuggestionClick} />
      );
    });

  handleOnUserInput = userInput => this.props.onUserInput(userInput);

  renderContent = props => {
    const {
      messages,
      selfIcon,
      showSelfAvatar,
      botIcon,
      categories,
      showCategories,
      onCategorySelect,
      determineTypeComp,
      onSuggestionClick
    } = props;
    return showCategories ? (
      <div className={styles.categoriesWrapper}>
        {categories.map(item => (
          <CategoryCard
            key={item.key}
            className={classNames(item.className, styles.categoryCard)}
            name={item.name}
            shortName={item.shortName}
            iconClass={item.iconClass}
            onClick={onCategorySelect}
            value={item.key}
            bgColor={item.bgColor}
            textColor={item.textColor}
            iconColor={item.iconColor}
          />
        ))}
      </div>
    ) : (
      <div className={styles.chat}>
        <div
          className={styles.chatMsgsContainer}
          ref={chatMsgsContainer =>
            (this.chatMsgsContainer = chatMsgsContainer)
          }
        >
          {this.renderMessages(
            messages,
            selfIcon,
            botIcon,
            determineTypeComp,
            showSelfAvatar,
            onSuggestionClick
          )}
          {this.props.showLoaderIcon && <Loader />}
        </div>
        <TextInputForm onUserInput={this.handleOnUserInput} />
      </div>
    );
  };

  render = () => {
    const {
      onMinimiseMaximise,
      chatWindowState,
      showCategories,
      onBack,
      appName,
      title
    } = this.props;
    setTimeout(() => {
      if (
        this.chatMsgsContainer &&
        this.chatMsgsContainer.scrollHeight >
          this.chatMsgsContainer.offsetHeight
      ) {
        this.chatMsgsContainer.scrollTop =
          this.chatMsgsContainer.scrollHeight -
          this.chatMsgsContainer.offsetHeight;
      }
    }, 0);
    return (
      <div className={styles.chatWrapper}>
        <Header
          title={!showCategories ? appName.shortName : title}
          onMinimise={onMinimiseMaximise}
          onMaximise={onMinimiseMaximise}
          onStretch={onMinimiseMaximise}
          onRestore={onMinimiseMaximise}
          isDocked={chatWindowState === 0}
          showBackIcon={!showCategories}
          onBack={onBack}
          windowState={chatWindowState}
        />
        {this.renderContent(this.props)}
      </div>
    );
  };
}

ChatWindow.displayName = 'ChatWindow';

ChatWindow.propTypes = {
  /** Handler for handling user input */
  onUserInput: PropTypes.func.isRequired,
  /** Chat messages including both bot response and user input */
  messages: PropTypes.array.isRequired,
  /** User's avatar icon */
  selfIcon: PropTypes.string.isRequired,
  /** Bot avatar icon */
  botIcon: PropTypes.string.isRequired,
  /** Handler to handle minimise, maximise & zoom of chat window */
  onMinimiseMaximise: PropTypes.func.isRequired,
  /** Handler for handling category selection */
  onCategorySelect: PropTypes.func.isRequired,
  /** Handler to going back to categories selection view from chatting section minimise of chat window */
  onBack: PropTypes.func.isRequired,
  /** Indicates the present state of chat window like minimise/maximise/zoom... */
  chatWindowState: PropTypes.oneOf([0, 1, 2]).isRequired,
  /** Title for the chat window when no categories are selected */
  title: PropTypes.string.isRequired,
  /** Name of the application/selected category name */
  appName: PropTypes.object,
  /** Indicates whether the zoom capability needs to be supported by the chat window */
  isZoomSupported: PropTypes.bool,
  /** Handler for handling suggestion selection from the chat response */
  onSuggestionClick: PropTypes.func,
  /** Function which determines the type of components to be used for the given response */
  determineTypeComp: PropTypes.func,
  /** Indicator for showing/hiding loader icon */
  showLoaderIcon: PropTypes.bool,
  /** Indicator for showing/hiding categories selection view */
  showCategories: PropTypes.bool,
  /** List of categories available/supported by the app  */
  categories: PropTypes.array,
  /** Indicator for showing/hiding user's self avatar icon */
  showSelfAvatar: PropTypes.bool
};

ChatWindow.defaultProps = {
  showLoaderIcon: false,
  showCategories: true,
  showSelfAvatar: false,
  isZoomSupported: false,
  categories: [],
  determineTypeComp: chatMsgConditionFunc,
  onSuggestionClick: () => {}
};

export default ChatWindow;
