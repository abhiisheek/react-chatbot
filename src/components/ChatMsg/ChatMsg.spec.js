/* global global */
import React from 'react';
import renderer from 'react-test-renderer';

import { shallow, mount } from 'enzyme';
import jsdom from 'jsdom';

import types from '../types';

import ChatMsg from './ChatMsg';

import { chatMsgConditionFunc } from '../ChatWindow/ChatMsgConditionalFunc';

// For mount
const doc = jsdom.jsdom('<!doctype html><html><body></body></html>');
global.document = doc;
global.window = doc.defaultView;

// Mock for snapshot testing
jest.mock('../utils/DateTime');

const props = {
  self: true,
  data: 'Hello',
  selfIcon: 'self',
  botIcon: 'bot',
  type: types.TEXT,
  showSelfAvatar: false,
  determineTypeComp: chatMsgConditionFunc
};

describe('ChatMsg', () => {
  it('Should have displayName set on the component', () => {
    expect(ChatMsg.displayName).toBeDefined();
  });

  it('Snapshot test', () => {
    const snapshot = renderer.create(<ChatMsg {...props} />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it('Should render the msg has self message when self is set', () => {
    const newProps = { ...props, self: true };
    const wrapper = shallow(<ChatMsg {...newProps} />);
    expect(wrapper.hasClass('selfWrapper')).toBeTruthy();
  });

  it('Should render the msg has bot message when self is unset', () => {
    const newProps = { ...props, self: false };
    const wrapper = shallow(<ChatMsg {...newProps} />);
    expect(wrapper.hasClass('selfWrapper')).toBeFalsy();
  });

  it('Should render the msg as text message when type is text', () => {
    const newProps = { ...props };
    const wrapper = mount(<ChatMsg {...newProps} />);
    expect(wrapper.find('.textMsg').exists()).toBeTruthy();
  });

  it('Should render the msg as text message when type is undefined/unset', () => {
    const newProps = { ...props, type: undefined };
    const wrapper = mount(<ChatMsg {...newProps} />);
    expect(wrapper.find('.textMsg').exists()).toBeTruthy();
  });

  it('Should render the msg as text message when type is not matched', () => {
    const newProps = { ...props, type: 'abc' };
    const wrapper = mount(<ChatMsg {...newProps} />);
    expect(wrapper.find('.textMsg').exists()).toBeTruthy();
  });

  it('Should render the self avatar when showSelfAvatar is set', () => {
    const newProps = { ...props, showSelfAvatar: true };
    const wrapper = mount(<ChatMsg {...newProps} />);
    expect(wrapper.find('.self.icon').exists()).toBeTruthy();
  });

  it('Should not render the self avatar when showSelfAvatar is set', () => {
    const newProps = { ...props, showSelfAvatar: false };
    const wrapper = mount(<ChatMsg {...newProps} />);
    expect(wrapper.find('.self.icon').exists()).toBeFalsy();
  });
});
