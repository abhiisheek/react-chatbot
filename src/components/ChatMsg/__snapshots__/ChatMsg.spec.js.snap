// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMsg Snapshot test 1`] = `
<div
  className="chatMsgWrapper selfWrapper"
>
  <div
    className="msg selfMsg"
  >
    <div
      className="chatMsgWrapper selfMsgWrapper"
    >
      <div
        className="textMsg wrapper"
      >
        Hello
      </div>
    </div>
  </div>
  <div
    className="timestamp"
  >
    10:11 PM
  </div>
</div>
`;
